soc/arm/microchip_mec/mec172x/soc.c

```diff
@@ -0,0 +1,411 @@
+		vbcs |= MCHP_VBATR_CS_XTAL_SE;
+	}
+
+	/* XTAL mode and enable must be separate writes */
+	vbr->CLK32_SRC = vbcs | MCHP_VBATR_CS_XTAL_CNTR_DG;
+	vbr->CLK32_SRC |= MCHP_VBATR_CS_SO_EN;
+	spin_delay(CLK32K_XTAL_WAIT);
+	vbr->CLK32_SRC |= MCHP_VBATR_CS_XTAL_DHC;
+
+	return 0;
+}
+
+/*
+ * Counter checks:
+ * 32KHz period counter minimum for pass/fail: 16-bit
+ * 32KHz period counter maximum for pass/fail: 16-bit
+ * 32KHz duty cycle variation max for pass/fail: 16-bit
+ * 32KHz valid count minimum: 8-bit
+ *
+ * 32768 Hz period is 30.518 us
+ * HW count resolution is 48 MHz.
+ * One 32KHz clock pulse = 1464.84 48 MHz counts.
+ */
+#define CNT32K_TMIN 1435
+#define CNT32K_TMAX 1495
+#define CNT32K_DUTY_MAX 74
+#define CNT32K_VAL_MIN 4
+
+static int check_32k_crystal(void)
+{
+	struct pcr_regs *pcr = PCR_XEC_REG_BASE;
+
+	pcr->CNT32K_CTRL = 0U;
+	pcr->CLK32K_MON_IEN = 0U;
+	pcr->CLK32K_MON_ISTS = MCHP_PCR_CLK32M_ISTS_MASK;
+
+	pcr->CNT32K_PER_MIN = CNT32K_TMIN;
+	pcr->CNT32K_PER_MAX = CNT32K_TMAX;
+	pcr->CNT32K_DV_MAX = CNT32K_DUTY_MAX;
+	pcr->CNT32K_VALID_MIN = CNT32K_VAL_MIN;
+
+	pcr->CNT32K_CTRL =
+		MCHP_PCR_CLK32M_CTRL_PER_EN | MCHP_PCR_CLK32M_CTRL_DC_EN |
+		MCHP_PCR_CLK32M_CTRL_VAL_EN | MCHP_PCR_CLK32M_CTRL_CLR_CNT;
+
+	uint32_t status = pcr->CLK32K_MON_ISTS;
+
+	while (status == 0) {
+		status = pcr->CLK32K_MON_ISTS;
+	}
+
+	pcr->CNT32K_CTRL = 0U;
+
+	if (status ==
+	    (MCHP_PCR_CLK32M_ISTS_PULSE_RDY | MCHP_PCR_CLK32M_ISTS_PASS_PER |
+	     MCHP_PCR_CLK32M_ISTS_PASS_DC | MCHP_PCR_CLK32M_ISTS_VALID)) {
+		return 0;
+	}
+
+	return -1;
+}
+
+static void connect_32k_source(enum clk32k_src src, enum clk32k_dest dest,
+			       uint32_t flags)
+{
+	struct pcr_regs *pcr = PCR_XEC_REG_BASE;
+	struct vbatr_regs *vbr = VBATR_XEC_REG_BASE;
+
+	if (dest == CLK32K_DEST_PLL) {
+		switch (src) {
+		case CLK32K_SRC_SIL_OSC:
+			pcr->CLK32K_SRC_VTR = MCHP_PCR_VTR_32K_SRC_SILOSC;
+			break;
+		case CLK32K_SRC_CRYSTAL:
+			pcr->CLK32K_SRC_VTR = MCHP_PCR_VTR_32K_SRC_SILOSC;
+			break;
+		case CLK32K_SRC_PIN:
+			pcr->CLK32K_SRC_VTR = MCHP_PCR_VTR_32K_SRC_PIN;
+			break;
+		default: /* do not touch HW */
+			break;
+		}
+	} else if (dest == CLK32K_DEST_PERIPH) {
+		uint32_t vbcs = vbr->CLK32_SRC & ~(MCHP_VBATR_CS_PCS_MSK);
+
+		switch (src) {
+		case CLK32K_SRC_SIL_OSC:
+			vbr->CLK32_SRC = vbcs | MCHP_VBATR_CS_PCS_VTR_VBAT_SO;
+			break;
+		case CLK32K_SRC_CRYSTAL:
+			vbr->CLK32_SRC = vbcs | MCHP_VBATR_CS_PCS_VTR_VBAT_XTAL;
+			break;
+		case CLK32K_SRC_PIN:
+			if (flags & CLK32K_FLAG_PIN_FB_CRYSTAL) {
+				vbr->CLK32_SRC =
+					vbcs | MCHP_VBATR_CS_PCS_VTR_PIN_SO;
+			} else {
+				vbr->CLK32_SRC =
+					vbcs | MCHP_VBATR_CS_PCS_VTR_PIN_XTAL;
+			}
+			break;
+		default: /* do not touch HW */
+			break;
+		}
+	}
+}
+
+static int pll_wait_lock(uint32_t wait_cnt)
+{
+	struct pcr_regs *regs = PCR_XEC_REG_BASE;
+
+	while (!(regs->OSC_ID & MCHP_PCR_OSC_ID_PLL_LOCK)) {
+		if (wait_cnt == 0) {
+			return -1;
```

Comment:
-ETIMEOUT instead of -1 ? just a minor comment.

Reply:
Hi @VenkatKotakonda  It's done